### PR TITLE
TICK-352 estimate hawkes sum exp kernel with varying baseline

### PR DESCRIPTION
**Pull request by Mbompr**

### CODE REVIEW

File: tick/optim/model/hawkes_fixed_sumexpkern_leastsq.py [hunk 1]
```diff
@@ -79,21 +94,42 @@ class ModelHawkesFixedSumExpKernLeastSq(ModelHawkes):
         "decays": {
             "writable": True,
             "cpp_setter": "set_decays"
-        }
+        },
+        "n_baselines": {
+            "writable": True,
+            "cpp_setter": "set_n_baselines"
+        },
+        "_period_length": {
```
Comment (vinther) [inline tick/optim/model/hawkes_fixed_sumexpkern_leastsq.py]: Maybe this should not be there

File: tick/optim/model/src/hawkes_fixed_sumexpkern_leastsq.cpp [hunk 2]
```diff
@@ -208,8 +237,18 @@ void ModelHawkesFixedSumExpKernLeastSq::compute_weights_i(const ulong i) {
...
       double decay_u = decays[u];
-      Dg_i[u] += 1 - cexp(-decay_u * (end_time - t_k_i));
-
+      ArrayDouble Dg_i_u = view_row(Dg_i, u);
+      for (ulong p = 0; p < n_baselines; ++p) {
+        ulong n_passed_periods = static_cast<ulong>(std::floor(t_k_i / period_length));
+        double lower = n_passed_periods * period_length + (p * period_length) / n_baselines;
+        while (lower < end_time) {
+          double lb = std::max(t_k_i, lower);
+          double upper = std::min(lower + period_length / n_baselines, end_time);
```
Comment (vinther) [inline tick/optim/model/src/hawkes_fixed_sumexpkern_leastsq.cpp]: const :D

File: tick/optim/model/src/hawkes_fixed_sumexpkern_leastsq.cpp [hunk 3]
```diff
@@ -254,5 +299,39 @@ void ModelHawkesFixedSumExpKernLeastSq::compute_weights() {
 }
 
 ulong ModelHawkesFixedSumExpKernLeastSq::get_n_coeffs() const {
-  return n_nodes + n_nodes * n_nodes * n_decays;
+  return n_nodes * n_baselines + n_nodes * n_nodes * n_decays;
+}
+
+ulong ModelHawkesFixedSumExpKernLeastSq::get_baseline_interval(const double t) {
+  double first_period_t = t - std::floor(t / period_length) * period_length;
```
Comment (vinther) [inline tick/optim/model/src/hawkes_fixed_sumexpkern_leastsq.cpp]: const

File: tick/optim/model/tests/hawkes_utils.py [hunk 4]
```diff
@@ -76,3 +76,27 @@ def hawkes_sumexp_kernel_intensities(baseline, decays, adjacency, timestamps):
...
+def hawkes_sumexp_kernel_varying_intensities(baseline, decays, adjacency,
+                                             timestamps):
+    # in this case baseline is a function of time
+    dim = len(timestamps)
+
+    kernels = {}
+    for i in range(dim):
+        kernels[i] = {}
+        for j in range(dim):
+            kernels[i][j] = lambda t, i=i, j=j: \
```
Comment (vinther) [inline tick/optim/model/tests/hawkes_utils.py]: This is perhaps better moved into a function declaration

File: tick/optim/model/src/hawkes_fixed_sumexpkern_leastsq.cpp (hunk 2 above)
Comment (Mbompr) [inline tick/optim/model/src/hawkes_fixed_sumexpkern_leastsq.cpp]: done

File: tick/optim/model/src/hawkes_fixed_sumexpkern_leastsq.cpp (hunk 3 above)
Comment (Mbompr) [inline tick/optim/model/src/hawkes_fixed_sumexpkern_leastsq.cpp]: done

File: tick/optim/model/src/hawkes_fixed_sumexpkern_leastsq.cpp [hunk 5]
```diff
@@ -254,5 +299,39 @@ void ModelHawkesFixedSumExpKernLeastSq::compute_weights() {
...
+  if (first_period_t == period_length) return n_baselines - 1;
+  return static_cast<ulong>(std::floor(first_period_t / period_length * n_baselines));
+}
+
+double ModelHawkesFixedSumExpKernLeastSq::get_baseline_interval_length(const ulong p) {
```
Comment (Mbompr) [inline tick/optim/model/src/hawkes_fixed_sumexpkern_leastsq.cpp]: changed to interval_p

File: tick/optim/model/hawkes_fixed_sumexpkern_leastsq.py (hunk 1 above)
Comment (Mbompr) [inline tick/optim/model/hawkes_fixed_sumexpkern_leastsq.py]: removed

File: tick/optim/model/tests/hawkes_utils.py (hunk 4 above)
Comment (Mbompr) [inline tick/optim/model/tests/hawkes_utils.py]: done

**Pull request merged**